api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -420,16 +431,71 @@ private User initializeUserIfNeeded() {
   }
 
   @Override
-  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest){
+  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest) {
     verifyInvitationKey(invitationVerificationRequest.getInvitationKey());
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
 
-  private void verifyInvitationKey(String invitationKey){
-    if(invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
+  private void verifyInvitationKey(String invitationKey) {
+    if (invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
       throw new BadRequestException(
-        "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+          "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+    }
+  }
+
+  /*
+   * This un-authed API method is limited such that we only allow contact email updates before the user has signed in
+   * with the newly created gsuite account. Once the user has logged in, they can change their contact email through
+   * the normal profile update process.
+   */
+  @Override
+  public ResponseEntity<Void> updateContactEmail(UpdateContactEmailRequest updateContactEmailRequest) {
+    String username = updateContactEmailRequest.getUsername();
+    com.google.api.services.admin.directory.model.User googleUser =
+      directoryService.getUser(username);
+    User user = userDao.findUserByEmail(username);
+    String newEmail = updateContactEmailRequest.getContactEmail();
+    if (userNeverLoggedIn(googleUser)) {
+      try {
+        new InternetAddress(newEmail).validate();
+      } catch (AddressException e) {
+        log.log(Level.INFO, "Invalid email entered.");
+        return ResponseEntity.badRequest().build();
+      }
+    } else {
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
     }
+    user.setContactEmail(newEmail);
+    return resetPasswordAndSendWelcomeEmail(username);
+  }
+
+  @Override
+  public ResponseEntity<Void> resendWelcomeEmail(ResendWelcomeEmailRequest resendRequest) {
+    String username = resendRequest.getUsername();
+    com.google.api.services.admin.directory.model.User googleUser =
+      directoryService.getUser(username);
+    User user = userDao.findUserByEmail(username);
+    if (userNeverLoggedIn(googleUser)) {
```

Comment:
nit: I would probably invert this and return early, and keep the normal codepath unindented.